### PR TITLE
feat: Effect.onComplete callback as an alternative to onFinish()

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

`Effect.onFinish` is now a property that can be directly set by the user. This allows an alternative way to chain multiple effects, and to perform some special action when an effect completes.

Previously, `Effect.onFinish()` was a method, and the only way to use it was to subclass a particular effect and create an overridden `.onFinish` method -- which is very cumbersome.

Note the asymmetry between `onStart()` and `onFinish()`: the first still remains an overridable method. This is because there are many effects that are interested in performing a custom action at start, and for them using a callback function would be more difficult. At the same time, there doesn't appear to be a clear use case when the user would want to provide a custom action at the start of an event.

### CODE REVIEW

File: packages/flame/lib/src/effects/effect.dart [hunk 1]
```diff
@@ -44,6 +44,15 @@ abstract class Effect extends Component {
   /// in order to make it run once again.
   bool removeOnFinish;
 
+  /// This method is called once when the effect is about to finish, but before
+  /// it is removed from parent. The notion of "about to finish" is defined by
```
Comment (spydon) [inline packages/flame/lib/src/effects/effect.dart:48]:
```suggestion
  /// it is removed from its parent. The notion of "about to finish" is defined by
```

**Review by spydon — APPROVED** [review]

**Review by erickzanardo — COMMENTED** [review]
We shouldn't introduce breaking changes with V1, an alternative here could probably be keep the method, receive a callback and the default implementation of the method, calls that callback.

**Comment by st-pasha** [comment]
> We shouldn't introduce breaking changes with V1, an alternative here could probably be keep the method, receive a callback and the default implementation of the method, calls that callback.

So, would you suggest that we mark the `onFinish` method as deprecated, or we'd rather always have both the method and the callback simultaneously?

**Comment by erickzanardo** [comment]
I would rather have the method and the callback.

File: doc/effects.md [hunk 2]
```diff
@@ -79,6 +79,9 @@ functionality inherited by all other effects. This includes:
     removed from the game tree and garbage-collected once the effect completes. Set this to false
     if you plan to reuse the effect after it is finished.
 
+  - Optional user-provided `onComplete` callback, which will be invoked when the effect has just
```
Comment (spydon) [inline doc/effects.md:82]: It is a bit confusing to have one function that is called `onComplete` and one that is called `onFinish`, not sure how to solve it though.

Comment (st-pasha) [reply doc/effects.md:82]: Yeah, I guess that's why Flutter style guide insists on naming convention "onComplete" for callbacks, and "handleComplete" for methods.

Comment (erickzanardo) [reply doc/effects.md:82]: We could just name it `onFinishCallback`

**Review by luanpotter — APPROVED** [review]
LGTM

**Review by spydon — APPROVED** [review]

**Pull request merged**